wgpu/gui/glfw.py

```diff
@@ -79,31 +55,71 @@ def __init__(self, *, size=None, title=None):
             if "wayland" in os.getenv("XDG_SESSION_TYPE", "").lower():
                 glfw.window_hint(glfw.FOCUSED, False)  # prevent Wayland focus error
 
-        self._window = glfw.create_window(width, height, title, None, None)
-        glfw.set_window_refresh_callback(self._window, self._paint)
+        # Create the window (the initial size may not be in logical pixels)
+        self._window = glfw.create_window(int(size[0]), int(size[1]), title, None, None)
+
+        # Register callbacks
+        glfw.set_window_content_scale_callback(self._window, self._on_pixelratio_change)
+        glfw.set_framebuffer_size_callback(self._window, self._on_size_change)
+        glfw.set_window_refresh_callback(self._window, self._on_paint)
         if sys.platform.startswith("darwin"):
             # Apparently, the refresh_callback is not called when the window
             # is created, gets focus, or is resized on macOS. So this is a
             # workaround to explicitely make sure that the paint callback
             # is called so that the contents are drawn.
             glfw.set_window_focus_callback(self._window, self._paint)
-            glfw.set_window_size_callback(self._window, self._paint)
+        # glfw.set_window_iconify_callback
+        # glfw.set_window_maximize_callback
+        # glfw.set_framebuffer_size_callback
 
-    def get_size_and_pixel_ratio(self):
-        # todo: maybe expose both logical size and physical size instead?
-        width, height = glfw.get_window_size(self._window)
-        pixelratio = glfw.get_window_content_scale(self._window)[0]
-        return width, height, pixelratio
+        # Initialize the size
+        self.set_logical_size(*size)
 
-    def get_display_id(self):
-        if sys.platform.startswith("linux"):
-            is_wayland = "wayland" in os.getenv("XDG_SESSION_TYPE", "").lower()
-            if is_wayland:
-                return glfw.get_wayland_display()
-            else:
-                return glfw.get_x11_display()
-        else:
-            raise RuntimeError(f"Cannot get GLFW display id on {sys.platform}.")
+    # Callbacks
+
+    def _on_pixelratio_change(self, *args):
+        self._set_logical_size()
+        self._on_paint()
+
+    def _on_size_change(self, *args):
+        self._logical_size = self._get_logical_size()
+        self._on_paint()
+
+    def _on_paint(self, *args):
+        self._draw_frame_and_present()
+
+    # Helpers
+
+    def _get_logical_size(self):
+        # Because the value of get_window_size is ambiguous, we use the
```

Comment:
ambiguous how? can you add that to the comment?

Reply:
updated.